MarvelApiClient/src/test/java/com/karumi/MarvelApiClient/CharacterApiClientTest.java

```diff
@@ -0,0 +1,251 @@
+        .addComic(ANY_COMIC_2)
+        .addEvents(getAnyEvents())
+        .addSerie(ANY_SERIE)
+        .addStory(ANY_STORY)
+        .build();
+    characterApiClient.getAll(query);
+
+    assertRequestSentToContains("offset=" + ANY_OFFSET, "limit=" + ANY_LIMIT, "name=" + ANY_NAME,
+        "nameStartsWith=" + ANY_START_NAME, "modifiedSince=" + ANY_MODIFIED_SINCE,
+        "orderBy=" + ORDER_NAME_DESCEDANT_VALUE, "comics=1,2", "events=1,2", "series=1",
+        "stories=1");
+  }
+
+  @Test(expected = MarvelApiException.class) public void shouldThrowExceptionWhenApiReturnAnError()
+      throws Exception {
+    CharacterApiClient characterApiClient = givenCharacterApiClient();
+    enqueueMockResponse(401,
+        "{\"code\":\"InvalidCredentials\",\"message\":\"That hash, timestamp and key "
+        + "combination is invalid.\"}");
+
+    try {
+      characterApiClient.getAll(ANY_OFFSET, ANY_LIMIT);
+    } catch (MarvelApiException e) {
+      assertEquals("InvalidCredentials", e.getMarvelCode());
+      assertEquals("That hash, timestamp and key combination is invalid.", e.getMessage());
+      throw e;
+    }
+  }
+
+  @Test public void shouldReturnAValidResponseWhenCallGetAll() throws Exception {
+    CharacterApiClient characterApiClient = givenCharacterApiClient();
+    enqueueMockResponse("getCharacters.json");
+
+    MarvelResponse<CharactersDto> characters = characterApiClient.getAll(0, ANY_LIMIT);
+
+    assertBasicMarvelResponse(characters);
+    CharactersDto charactersDto = characters.getResponse();
+    assertEquals(10, charactersDto.getCount());
+    assertEquals(10, charactersDto.getLimit());
+    assertEquals(0, charactersDto.getOffset());
+    assertEquals(1485, charactersDto.getTotal());
+
+    CharacterDto firstCharacter = charactersDto.getCharacters().get(0);
+    assertIs3dManCharacter(firstCharacter);
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void shouldReturnAnIllegalExceptionWhenCallGetCharacterWithoutId() throws Exception {
+    CharacterApiClient characterApiClient = givenCharacterApiClient();
+
+    characterApiClient.getCharacter(INVALID_CHARACTER_ID);
+  }
+
+  @Test public void shouldReturnAValidUrlWhenGetCharacterHasAValidId() throws Exception {
+    CharacterApiClient characterApiClient = givenCharacterApiClient();
+    enqueueMockResponse("getCharacter.json");
+
+    characterApiClient.getCharacter(THREED_MAN_CHARACTER_ID);
+
+    assertRequestSentToContains("characters/" + THREED_MAN_CHARACTER_ID);
+  }
+
+  @Test(expected = MarvelApiException.class)
+  public void shouldReturnMarvelExceptionWhenTheIdDoesNotExist() throws Exception {
+    CharacterApiClient characterApiClient = givenCharacterApiClient();
+    enqueueMockResponse(404, "{\"code\":404,\"status\":\"We couldn't find that character\"}");
+
+    try {
+      characterApiClient.getCharacter(ANY_NOT_FOUND_ID);
+    } catch (MarvelApiException e) {
+      assertEquals("404", e.getMarvelCode());
```

Comment:
exist 2 codes, the httpCode and the MarvelCode, that it's an internal error that have marvel for his requests, sometimes httpCode and the MarvelCode could be different, and I called Marvel Code for we can easy diferenciate the httpResponseCode and the marvelCode.